core/drivers/firewall/stm32_risab.c

```diff
@@ -0,0 +1,843 @@
+	if (rif_conf & BIT(RISAB_DPRIV_SHIFT)) {
+		subr_cfg->dprivcfgr = _RISAB_PG_PRIVCFGR_MASK;
+		if (check_overlap &&
+		    reg_pages_cfged & risab_d->pages_configured)
+			panic("Memory region overlap detected");
+	} else {
+		subr_cfg->dprivcfgr = 0;
+	}
+
+	if (check_overlap)
+		risab_d->pages_configured |= reg_pages_cfged;
+
+	for (i = 0; i < _RISAB_NB_MAX_CID_SUPPORTED; i++) {
+		/* RISAB compartment priv configuration */
+		if (rif_conf & BIT(i)) {
+			subr_cfg->plist[i] |=
+			GENMASK_32(first_page +
+				   subr_cfg->nb_pages_cfged - 1,
+				   first_page);
+		}
+
+		/* RISAB compartment read configuration */
+		if (rif_conf & BIT(i + RISAB_READ_LIST_SHIFT)) {
+			subr_cfg->rlist[i] |=
+			GENMASK_32(first_page +
+				   subr_cfg->nb_pages_cfged - 1,
+				   first_page);
+		}
+
+		/* RISAB compartment write configuration */
+		if (rif_conf & BIT(i + RISAB_WRITE_LIST_SHIFT)) {
+			subr_cfg->wlist[i] |=
+			GENMASK_32(first_page +
+				   subr_cfg->nb_pages_cfged - 1,
+				   first_page);
+		}
+	}
+
+	/* CID filtering configuration */
+	if (rif_conf & BIT(RISAB_CFEN_SHIFT))
+		subr_cfg->cidcfgr |= _RISAB_PG_CIDCFGR_CFEN;
+
+	if (rif_conf & BIT(RISAB_DCEN_SHIFT))
+		subr_cfg->cidcfgr |= _RISAB_PG_CIDCFGR_DCEN;
+
+	if (rif_conf & RISAB_DCCID_MASK) {
+		uint32_t ddcid = SHIFT_U32((rif_conf & RISAB_DCCID_MASK) >>
+					   RISAB_DCCID_SHIFT,
+					   _RISAB_PG_CIDCFGR_DDCID_SHIFT);
+
+		assert(((rif_conf & RISAB_DCCID_MASK) >> RISAB_DCCID_SHIFT) <
+		       _RISAB_NB_MAX_CID_SUPPORTED);
+
+		subr_cfg->cidcfgr |= ddcid;
+	}
+}
+
+static TEE_Result parse_dt(const void *fdt, int node,
+			   struct stm32_risab_pdata *risab_d)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	const fdt32_t *mem_regions = NULL;
+	struct dt_node_info info = { };
+	const fdt32_t *cuint = NULL;
+	int mem_reg_node = 0;
+	unsigned int i = 0;
+	int lenp = 0;
+
+	fdt_fill_device_info(fdt, &info, node);
+	assert(info.reg != DT_INFO_INVALID_REG &&
+	       info.reg_size != DT_INFO_INVALID_REG_SIZE);
+
+	risab_d->base.pa = info.reg;
+
+	/* Gate the IP */
+	res = clk_dt_get_by_index(fdt, node, 0, &risab_d->clock);
+	if (res)
+		return res;
+
+	strlcpy(risab_d->risab_name, fdt_get_name(fdt, node, NULL),
+		sizeof(risab_d->risab_name));
+
+	cuint = fdt_getprop(fdt, node, "st,srwiad", NULL);
+	if (cuint)
+		risab_d->srwiad = true;
+
+	/* Get the memory region being configured */
+	cuint = fdt_getprop(fdt, node, "st,mem-map", &lenp);
+	if (!cuint)
+		panic("Missing st,mem-map property in configure memory region");
+
+	assert((unsigned int)(lenp / sizeof(uint32_t)) == 2);
+
+	risab_d->region_cfged.base = fdt32_to_cpu(cuint[0]);
+	risab_d->region_cfged.size = fdt32_to_cpu(cuint[1]);
+
+	/* Get the memory regions to configure */
+	mem_regions = fdt_getprop(fdt, node, "memory-region", &lenp);
+	if (!mem_regions)
+		panic("No memory region to configure");
+
+	risab_d->nb_regions_cfged = (unsigned int)(lenp / sizeof(uint32_t));
+	assert(risab_d->nb_regions_cfged < _RISAB_NB_PAGES_MAX);
```

Comment:
could you assert `risab_d->nb_regions_cfged > 0` to ensure all `subr_cfg->nb_pages_cfged - 1` occurrences are reliable.